### ISSUE TITLE
Implement merging multiple documents into one

### ISSUE THREAD

**Issue by Lvaskz** [junrrein/pdfslicer#30]
Hi, this tool is great, but it would be greater if we can the posibility to join several pdf files into one, as a new feature.

**Comment by junrrein** [comment]
I was expecting someone to suggest this.

I'm not opposed to this feature, but implementing this will require rethinking a lot of the current UI. It would be really easy to add this feature in an a way that feels unergonomic and awkward.

Since I'm not good at UI design, I'll we waiting for someone else to come up with a design that shows how this feature would work in the context of the current UI. If no good design comes up, this won't get implemented.

**Comment by sojusnik** [comment]
Why not by opening several pdfs at once? Or allowing to drag and drop pdfs onto the main window? In this way we could intuitively control the order of the pdfs and therefore merge several into one.

Also shuffling pages would be very handy by dragging and dropping them around.

**Comment by junrrein** [comment]
Opening several PDFs at once could work. Allowing to drag and drop pdfs onto the main window is also something that could work. Reordering pages using drag and drop could also work.

My concern is that every action that can be done using drag and drop should also be able to be done using buttons in the UI. And the UI has more than enough buttons already, so some kind of design work is necessary here.

**Comment by junrrein** [comment]
I have a small doubt about the UI for this.

Currently, Slicer shows the filename of the opened PDF as a subtitle on the headerbar. When there are multiple opened files, having this subtitle won't make sense.

I have a couple ideas in mind:

1. Use the approach taken by [PDF-Shuffler](https://a.fsdn.com/con/app/proj/pdfshuffler/screenshots/181783.jpg/max/max/1) all the time.

2. When there is only 1 open file, show subtitle on headerbar. When there are more, use approach (1).

What do you think?

(tagging @amivaleo, who usually likes to comment on UI issues).

**Comment by amivaleo** [comment]
> (tagging @amivaleo, who usually likes to comment on UI issues)
😄

Well...
I have several UI ideas in mind. The main issue, to me, is the page-grid itself, as I already told you, Julian, here #68.

So:

- the page-grid should be very similar to that used by nautilus, on gnome-boxed (in grid mode), or gnome-games-app, or gnome-music, etc. Meaning: the selected page should have a blue background just like a selected file in nautilus.
- the selection method should be exactly the same as that used by nautilus -> this sets to 0 the effort required for the user in order to learn how to select pages in Slicer and, as a bonus, sets to 0 the effort of Julian who won't need to reinvent the wheel! :)
- the print-button on each page should be removed. It triggers the action "show me a preview". I believe that such action should be triggered by a text-based button on the bottom toolbar. This button should only appear when *only one* page is selected. It should be hidden (not only grayed out). The action itself should be triggered by pressing the space key (just like in nautilus...).
- remove the couplet of zoom button -/+ and put a slider in the hamburger menu for zooming.
- transform the icon-based "open" button into a text-based.
- move the save-button to the right and make it text-based.
- add a "save as..." icon-based button next (and attached) to it. This should be the final result (as in glade):
![Schermata da 2019-05-03 16-16-23](https://user-images.githubusercontent.com/24316547/57143279-d10cc680-6dbe-11e9-9cd4-ea31772c79f4.png)
- remove the title "pdf slicer" from the headerbar and keep only the open pdf name as only-&-main title.
- add an asterisk ( * ) at the beginning of the title when the open pdf is edited, like other apps do (gedit, gimp).
- add a "+" button to the right of the open button, which is grayed out only when no pdf is open. That button opens a file selector -> the user chooses a pdf -> the new pdf is ADDED the open one, at the end of it and ALL of its pages are automatically selected. So that the user can click on left/right buttons to move all the new pages, or drag&drop them, or rotate them, whatever.
- IF others pdf files are added to the original one, the title should show a subtitle "added file NAME_FILE.PDF" if only one file is added, or "multiple files added" if more than one file is added. In this way, if the user click on the "save" button, he will overwrites the original pdf.
- IF other pdf files are added, every page should show a two-rows label. The top row should be the name of the file that page comes from. The second/bottom row should be the original page number. So that the user could recognize both the origin every page and its number.
- the "add-blank-page" button, if/when you will add it, should stay on the bottom toolbar, in order not to mix what you can do on a "file-point-of-view" (topbar/headerbar) from what you can do on a "page-point-of-view" (bottombar/toolbar).

It's a bunch of changes. I am sure you won't like all of them. I definitely hope you will argue on many of these, indeed, since I'm surely forgetting something! :D

**Comment by junrrein** [comment]
@amivaleo 

I'm loving all of these suggestions you've made, even if I don't agree with them all (but I agree with most!).

I'll be creating separate issues for each aspect you are talking about that is not directly relevant to this issue, tag you, and make my own comments.

**Comment by junrrein** [comment]
@amivaleo 

> add a "save as..." icon-based button next (and attached) to it. This should be the final result (as in glade):
> ![](https://user-images.githubusercontent.com/24316547/57143279-d10cc680-6dbe-11e9-9cd4-ea31772c79f4.png)

This is a direction I specifically don't want to take.

Since removing pages is one of the primary functions of the app (and my primary use case!) I want to make it hard to overwrite opened files, and thus losing pages forever. Having only a `Save as...` button accomplishes this, since you have to select the specific original file if you wish to overwrite it.

**Comment by junrrein** [comment]
@amivaleo 

> the "add-blank-page" button, if/when you will add it, should stay on the bottom toolbar, in order not to mix what you can do on a "file-point-of-view" (topbar/headerbar) from what you can do on a "page-point-of-view" (bottombar/toolbar).

I agree completely, I have nothing to add here.

**Comment by amivaleo** [comment]
> I'm loving all of these suggestions you've made, even if I don't agree with them all (but I agree with most!).

I'm glad to see that you appreciate. :) This is a bunch of ideas I had in my mind but I wasn't sure about telling them to you. Sometimes I wrote my opinions and had the feeling that I was *bothering* the dev. :\

> I want to make it hard to overwrite opened files

Ok, I missed that. It makes sense, indeed. Stick with that. :)

> any metadata coming from the original file is not copied over

Gosh, this is a weak point for me. I miss a gui-based metadata editor for pdf.

I understand your choice of "creating a new file" instead of editing the open one. Although it opens an issue: what if I only need to rotate some page or add a few blank pages? I would lose the table of contents and all the metadata. :\

> All of these are good points, but I think all of it may depend on where we go with the previous aspect, so we can talk about these later.

💪

**Comment by junrrein** [comment]
@amivaleo 

> I'm glad to see that you appreciate. :) This is a bunch of ideas I had in my mind but I wasn't sure about telling them to you. Sometimes I wrote my opinions and had the feeling that I was bothering the dev. :\

I'm not very comfortable in the design department, so any and all suggestions made in this area are welcome.

> I understand your choice of "creating a new file" instead of editing the open one. Although it opens an issue: what if I only need to rotate some page or add a few blank pages? I would lose the table of contents and all the metadata. :\

Yeah, the more I'm told about this, the more I can see it for the issue it is. Sometime in the future I'll try to keep the metadata on saved documents.

Let's summarize what we've agreed on:

* There'll be an `Open` button (text-labeled). Documents opened this way will replace whatever was already in the viewing area. There should be a confirmation dialog if there were unsaved modifications, right?
* There'll be an `Add document/file` button (icon-labeled). This icon will be placed next to the previous one. What should the tooltip of this button be? "Add document" or "Add file"? I'm thinking there should be a choice of inserting the new pages at the beginning or at the end of the document. Moving pages all that way will be cumbersome in large documents, even when the app gets drag-and-drop support.

**Comment by amivaleo** [comment]
> There'll be an Open button (text-labeled). Documents opened this way will replace whatever was already in the viewing area. There should be a confirmation dialog if there were unsaved modifications, right?

Yes, perfect.

> There'll be an Add document/file button (icon-labeled). This icon will be placed next to the previous one. What should the tooltip of this button be? "Add document" or "Add file"? I'm thinking there should be a choice of inserting the new pages at the beginning or at the end of the document. Moving pages all that way will be cumbersome in large documents, even when the app gets drag-and-drop support.

My suggestion is to simply use a "+" icon (list-add-symbolic, I suggest) for that button (that's way it's good to do something for the current zoom buttons, otherwise there would be two "+" icons, only slightly different.

Regarding the "where to add the new pages from that file", I believe you should have a way to some custom content on the "open file" widget.

Consider that widget in gedit and libreoffice:

![Schermata da 2019-05-09 09-10-28](https://user-images.githubusercontent.com/24316547/57434055-51f21500-723a-11e9-9bb9-18d8b55c5025.png)

As you can see, the bottom bars are not the same for both applications, although the main widget is surely the same.

Can you do something similar for Slicer? In that case, I'd suggest to add a label and a dropdown button with only "left" and "right" as items.

**Comment by junrrein** [comment]
@amivaleo 

> My suggestion is to simply use a "+" icon (list-add-symbolic, I suggest) for that button (that's way it's good to do something for the current zoom buttons, otherwise there would be two "+" icons, only slightly different.

I was thinking about using `list-add-symbolic` for adding blank pages, and `document-new-symbolic` for adding a file. Or maybe the other way around?

> Regarding the "where to add the new pages from that file", I believe you should have a way to some custom content on the "open file" widget.

I can't do this _and_ have it working with the current Flatpak permissions. Currently Gtk provides two kind of file pickers:

* One which you can customize as you've shown, but isn't sandboxed. This means the app will need access to the entire filesystem.
* The second one which you can't customize, and that integrates with the Flatpak sandbox so only files that the user has explicitly selected can be seen by the app.

Of course, I think usability probably should come first, so if push comes to save, we'll go with the "insecure" option.

The other option, using the permissions we currently have, is to have the dropdown menu on the "Add document" button itself.

By the way, what about the tooltip text for the button? "Add document" or "Add file"?

**Comment by amivaleo** [comment]
> Or maybe the other way around?

The document-new-symbolic is indeed good for both purposes *but* other applications use it for adding files, like: glade, glatex (well... It uses colorful icons 
![Schermata da 2019-05-09 16-46-05](https://user-images.githubusercontent.com/24316547/57462873-f8113f80-7279-11e9-840d-3350b2b215cb.png)
, but well...).

So yes, it is better to use for adding a file.

For the label: I checked Evince. It uses the label "Open Document" when you click on the "Open" button. Your Slicer handles PDF too, so... I'd go for "Add Document". :+1: 

One problem is left: the + button has some sense if it's close to the open-button, because it suggests that it does something similar to some extent. But if it goes on the bottombar (where it has to go if its purpose is to add a page), it becomes hard to understand what it does from a first glance... 🤔

"Add Blank Page" is quite a long a label for a text-button...

Maybe "New Page"? Or only "Add Page"...?

!

What if you make it a text-button with "Add Blank Page" BUT you make it appear only if ONE page is selected?

Just an idea, it could be weird.

**Comment by amivaleo** [comment]
> I can't do this and have it working with the current Flatpak permissions.

:(

> Moving pages all that way will be cumbersome in large documents, even when the app gets drag-and-drop support.

Yes, *moving* become cumbersome. But what about cut&paste? :)

**Comment by junrrein** [comment]
@amivaleo 

> One problem is left: the + button has some sense if it's close to the open-button, because it suggests that it does something similar to some extent. But if it goes on the bottombar (where it has to go if its purpose is to add a page), it becomes hard to understand what it does from a first glance... :thinking:

That is indeed true. But let's worry about the "Add blank page" button later at #60. 

Right now let's think about which is the icon that makes the most sense for the issue at hand. I do think both `list-add` and `document-new` make sense. But which makes the most sense?

`document-new` is usually used to create a new blank document in other apps (like glade and gnome-latex), while `list-add` may make more sense since we are adding to the current document.

> For the label: I checked Evince. It uses the label "Open Document" when you click the "Open" button. Your Slicer handles PDF too, so... I'd go for "Add Document". :+1:

:+1:

>> I can't do this and have it working with the current Flatpak permissions.
>
> :(

I'm thinking of three alternatives here:

1. Ditch the sandbox and provide a customized "Open file dialog".
2. Open a dialog where the user can select the file and insertion position. This is a rough sketch:

    ![Captura de pantalla de 2019-05-09 18-00-33](https://user-images.githubusercontent.com/962049/57486509-7e328380-7284-11e9-86ca-4d9271a7ac2f.png)

3. Have a dropdown on the "Add Document" button, which shows a menu with the different insertion positions. Upon clicking on either of them, the Open File dialog shows up.

**Comment by amivaleo** [comment]
> But which makes the most sense?

Gnome-games has a + button for adding games, which can be seen as a button that adds a file. Slicer is going to do the same, so...

My thinking:
1. "+" button for adding files close to the "open" button & "document-new" button on the bottom bar for adding pages: this is, I think, a clear and neat solution. Both icons would serve their purposes.
2. "document-new" button for adding files close to the "open" button & "+" button on the bottom bar for adding pages: the latter button is hard to understand. Although it's usually used for adding pages/things to the current open file, it should be placed at the end of the list of "things" (indeed, its name is "list-add").

I'm a bit confused by now. 😄

My solution has always been 1. It's quite logical, I think. I tried to see things in a different way in my previous post. Maybe there was no need for that.

...

My suggestion: 1.


On "how the add-file feature should work":

> Ditch the sandbox and provide a customized "Open file dialog".

This is a no-no solution. From some philosophical point of view: designing the gui should not affect security or the delivery-procedure or whatever. The gui is only the interface between the code and the user, it should not affect the quality or put limitations on the code in any way.

> Have a dropdown on the "Add Document" button, which shows a menu with the different insertion positions. Upon clicking on either of them, the Open File dialog shows up.

Forgive me, I think it's a bad design. :) I'm not an expert on this matter, just a man who loves his own OCD. 😂

> Open a dialog where the user can select the file and insertion position.

This might be a solution, but still too much effort. Before starting to follow this path, I want your opinion on this:

> Yes, moving become cumbersome. But what about cut&paste? :)

What I meant is that the user could:
1. click on the "add-document" button (will it be a "+"?)
2. choose the file he wants to add from a dialog window -> the new-pages are loaded after the old-pages, AND (this is crucial) the new-pages take presidence since they all get selected
3. the user can cut/ctrl+x the new-pages
4. select one (or several) old-page
5. paste/ctrl+v the new-pages
6. the new-pages will be added after the last selected old-page

This is harder to say than to do, I think.

An extra: if at least one of the old-pages is selected, the new-pages will be loaded after the last selected old-page.

I hope it's clear.

_Basically the idea is the user chooses where the add the new document according to which, of the current pages, he selects._

If this solution seems reasonable for you, I suggest to say more on the hint-message on the add-document button: "Add a document after the selected page(s)".

**Comment by junrrein** [comment]
> > But which makes the most sense?
> 
> Gnome-games has a + button for adding games, which can be seen as a button that adds a file. Slicer is going to do the same, so...
> 
> My thinking:
> 
>     1. "+" button for adding files close to the "open" button & "document-new" button on the bottom bar for adding pages: this is, I think, a clear and neat solution. Both icons would serve their purposes.
> 
>     2. "document-new" button for adding files close to the "open" button & "+" button on the bottom bar for adding pages: the latter button is hard to understand. Although it's usually used for adding pages/things to the current open file, it should be placed at the end of the list of "things" (indeed, its name is "list-add").
> 
> 
> I'm a bit confused by now. smile
> 
> My solution has always been 1. It's quite logical, I think. I tried to see things in a different way in my previous post. Maybe there was no need for that.
> 
> ...
> 
> My suggestion: 1.

I also like 1 more, let's go with that for now.

> On "how the add-file feature should work":
> 
> > Ditch the sandbox and provide a customized "Open file dialog".
> 
> This is a no-no solution. From some philosophical point of view: designing the gui should not affect security or the delivery-procedure or whatever. The gui is only the interface between the code and the user, it should not affect the quality or put limitations on the code in any way.

Well, philosophically, I think it should be the other way around: the way code works shouldn't impose any limitations on the GUI you want to achieve. But I completely understand why the "restrictive" Gtk file dialog works this way.

> > Have a dropdown on the "Add Document" button, which shows a menu with the different insertion positions. Upon clicking on either of them, the Open File dialog shows up.
> 
> Forgive me, I think it's a bad design. :) I'm not an expert on this matter, just a man who loves his own OCD. :joy:

Good to know.

> > Open a dialog where the user can select the file and insertion position.
> 
> This might be a solution, but still too much effort. Before starting to follow this path, I want your opinion on this:
> 
> > Yes, moving become cumbersome. But what about cut&paste? :)
> 
> What I meant is that the user could:
> 
>     1. click on the "add-document" button (will it be a "+"?)
> 
>     2. choose the file he wants to add from a dialog window -> the new-pages are loaded after the old-pages, AND (this is crucial) the new-pages take presidence since they all get selected
> 
>     3. the user can cut/ctrl+x the new-pages
> 
>     4. select one (or several) old-page
> 
>     5. paste/ctrl+v the new-pages
> 
>     6. the new-pages will be added after the last selected old-page
> 
> 
> This is harder to say than to do, I think.

Oh I forgot about cut & paste. It will work, but I think it's too cumbersome. And I don't know when I will get around to implement cut & paste, so I want to have either of the 3 options even if only as a stop-gap measure.

> An extra: if at least one of the old-pages is selected, the new-pages will be loaded after the last selected old-page.
> 
> I hope it's clear.
> 
> _Basically the idea is the user chooses where the add the new document according to which, of the current pages, he selects._
> 
> If this solution seems reasonable for you, I suggest to say more on the hint-message on the add-document button: "Add a document after the selected page(s)".

There are a couple of things I don't like about this solution:

* What will the behaviour of this button be when there are no selected pages? Stay disabled? This doesn't make sense. Should pages be added at the end when none are selected? Then the same button would have different behaviours depending on the context, which I don't like.
* This behaviour will be surprising on touch devices, since tooltips can't be seen there (though this is a minor thing for me).

The thing I'm liking the most right now is implementing the "dialog" solution (number 2), and then iterate on this once Slicer gets cut & paste and drag & drop.

**Comment by amivaleo** [comment]
> Good to know.

🤨

> Well, philosophically, I think it should be the other way around: the way code works shouldn't impose any limitations on the GUI you want to achieve. But I completely understand why the "restrictive" Gtk file dialog works this way.

This is against multiplatform. Maybe you're not planning to deliver Slicer on other OSes, but maybe one day there will be the need to change its GUI (for whatever reason) and then... Dunno, I like the "use less popups as possible".


> Should pages be added at the end when none are selected?

Well... ^_^

> Then the same button would have different behaviours depending on the context

Yep.

> which I don't like

:(

> I'm liking the most right now is implementing the "dialog" solution (number 2)

Listen to this other idea (please, appreciate my efforts in order to restrain you from this dialog-solution :D ):

1. click on the + button
2. the open-dialog appears -> select the file -> close that dialog
3. a blue banner appears on the main window (the same blue banner that you can find in nautilus -> Public folder) telling the user to select the page after which the new-pages will be added

**Comment by junrrein** [comment]
> > Good to know.
> 
> 🤨

I was just acknowledging your opinion, not trying to be sarcastic or anything.

> Listen to this other idea (please, appreciate my efforts in order to restrain you from this dialog-solution :D ):
> 
>     1. click on the + button
> 
>     2. the open-dialog appears -> select the file -> close that dialog
> 
>     3. a blue banner appears on the main window (the same blue banner that you can find in nautilus -> Public folder) telling the user to select the page after which the new-pages will be added

Let me tell that you that I really appreciate having you participate this way :)

With that being said, [Info Bars](https://developer.gnome.org/gtk3/stable/GtkInfoBar.html#GtkInfoBar.description) (the blue banner you are talking about) are meant for showing contextual information, or to show an action that can be made in that context (like the ones you can find in Nautilus -> Trash), not to require an user to do something.

I know this may look like me being a jerk, but I will go with one of the ideas I had in mind (number 3 in fact, "dropdown menu on the Add Document button"). It will be the easiest to code, and good enough for me. I may be open to revisit this design in the future, but it will do for now.

Again, sorry if it seems I am just brushing your ideas off, but it's not the case. It's just that we disagree on this (at least for now!). I hope this doesn't turn you off from contributing more ideas :sweat_smile:

**Comment by amivaleo** [comment]
> not trying to be sarcastic

You should have! :) 

No worries, Julian. I only get pi#sed off when people really want to pi#s me off.

> not to require an user to do something.

Indeed... I think that I once knew that, then I forgot it. I should re-read the HIGs again. It's been a while already.

> I hope this doesn't turn you off from contributing more ideas

Not at all.

I am not a devote fun of gnome overall gui (mainly: headerbars with hamburger menus). I like android approach: sidebars. But I like homogeneus UX above all. So, no problem if I get a no.

But ehi: if you entirely discard the mockup for which I spent weeks of my time, as if it where cr*p, then calmly

and peacefully,

using a very polite language

and without being rude,

I tell you "f#ck off, mo#on". 😂

**Comment by junrrein** [comment]
This is currently implemented in `master`, and will get included in the next release.

Thanks everyone for the input!
